src/kpi/statistics/developerSpread/developerSpread.service.ts

```diff
@@ -1,20 +1,138 @@
 import { Injectable, Logger } from '@nestjs/common';
-import { UserService } from '../../../entities/users/user.service';
-import { CommitService } from '../../../entities/commits/commit.service';
+import { RepositoryService } from '../../../entities/repositories/repository.service';
+import {
+  Intervals,
+  groupByIntervalSelector,
+  serialize,
+} from '../../statistics/lib';
 
 @Injectable()
 export class DeveloperSpreadService {
   private readonly logger = new Logger(DeveloperSpreadService.name);
 
-  constructor(
-    private readonly userService: UserService,
-    private readonly commitService: CommitService,
-  ) {}
+  constructor(private readonly repoService: RepositoryService) {}
 
-  async developerSpread() {
-    this.logger.log('Developer spread calculation');
-    for await (const user of this.userService.readAll()) {
-      console.log(user);
+  async developerSpread(
+    interval: Intervals = Intervals.MONTH,
+    owner: string,
+    repo: string = undefined,
+    since: string = undefined,
+    to: string = undefined,
+  ) {
+    const pipeline = this.repoService.preAggregate(
+      { owner: owner },
+      {
+        issues: { events: { actor: true, since: since, to: to } },
+        commits: { author: true, since: since, to: to },
+      },
+    );
+    pipeline.unwind('$commits');
+    pipeline.group({
+      _id: '$_id',
+      contributors: {
+        $addToSet: '$commits.author',
+      },
+      owner: { $first: '$owner' },
+      repo: { $first: '$repo' },
+      commits: { $push: '$commits' },
+      issues: { $first: '$issues' },
+    });
+    if (repo) {
+      pipeline.addFields({
+        contributors: {
+          $cond: {
+            if: { $eq: ['$repo', repo] },
+            then: '$contributors',
+            else: null,
+          },
+        },
+      });
+    }
+    pipeline.unwind('$commits');
+    pipeline.unwind('$issues');
+    pipeline.unwind('$issues.events');
+    pipeline.group({
+      _id: '$owner',
+      contributors: {
+        $addToSet: '$contributors',
+      },
+      commits: {
+        $addToSet: {
+          user: '$commits.author',
+          timestamp: '$commits.timestamp',
+          repo: '$repo',
+        },
+      },
+      events: {
+        $addToSet: {
+          user: '$issues.events.actor',
+          timestamp: '$issues.events.created_at',
+          repo: '$repo',
+        },
+      },
+    });
+    pipeline.project({
+      activities: {
+        $setUnion: ['$commits', '$events'],
+      },
+      contributors: 1,
+    });
+    // without a repo filter, contributors looks like:
+    //    [ [<contributors of repo 1>], [<contributors of repo 2>], [<contributors of repo 3>]]
+    // with a repo filter, contributors looks like:
+    //    [ [], [<contributors of filtered repo>]]
+    // I couldn't find a more elegant way to flatten this array than this combination of unwind-unwind-group
```

Comment:
just divide it into two separate methods? then we could also just query the one specific repository we are interested in? I actually don't understand how this works right now if we get the repo filter. Because the initial match in the pre aggregation method is done with just the owner. But if we now right from the beginning that we are just interested in a specific repository we could utilise this information right in that query?